packages/backend/src/peripherals/database/PageRepository.ts

```diff
@@ -0,0 +1,93 @@
+import { Knex } from 'knex'
+import { PageRow } from 'knex/types/tables'
+import { sortBy } from 'lodash'
+
+import { Logger } from '../../tools/Logger'
+import { Repository } from './types'
+
+export interface PageRecord {
+  id?: number
+  blockNumber: number
+  pageHash: string
+  data: string
+}
+
+export class PageRepository implements Repository<PageRecord> {
+  constructor(private knex: Knex, private logger: Logger) {
+    this.logger = logger.for(this)
+  }
+
+  async add(records: PageRecord[]) {
+    if (records.length === 0) {
+      this.logger.debug({ method: 'addOrUpdate', rows: 0 })
+      return
+    }
+
+    const rows: PageRow[] = records.map(toRow)
+    await this.knex('pages').insert(rows)
+
+    this.logger.debug({ method: 'add', rows: rows.length })
+  }
+
+  async getAll() {
+    const rows = await this.knex('pages').select('*')
+    this.logger.debug({ method: 'getAll', rows: rows.length })
+    return rows.map(toRecord)
+  }
+
+  async getAllForFacts(factHashes: string[]) {
+    type Row = { fact_hash: string; pages: string[] }
+
+    const rows = (await this.knex('fact_to_pages')
+      .select(
+        'fact_hash',
+        this.knex.raw(
```

Comment:
Ah the beauty of raw SQL. I would have done it in JS unaware of the possibilities.

Reply:
I keep thinking about putting this in a string array from the get go instead of splitting it on insertion and aggregating it back on read.